### PR TITLE
🌵 new compose component

### PR DESCRIPTION
**Pull request by susanm74**

### CODE REVIEW

**Comment by susanm74** [comment]
NB - chatbox charcount and unit tests still in progress - please don't merge yet!

File: src/contacts/ContactChat.ts [hunk 1]
```diff
@@ -292,25 +302,42 @@ export class ContactChat extends ContactStoreElement {
...
+    if (buttonName === 'Send') {
+      // add contact uuid
+      const payload = {
+        contacts: [this.currentContact.uuid],
+      };
+      // add translations text, attachments, etc
+      const compose = evt.currentTarget as Compose;
+      if (compose && (compose.currentChat || compose.values)) {
+        const translations = {
+          und: {
```
Comment (susanm74) [inline src/contacts/ContactChat.ts:317]: @rowanseymour do we prefer the translations objects to use the "und" (undetermined) vs "eng" (english) vs use whatever the base language is? wanted to confirm based on your notes here https://github.com/nyaruka/rapidpro/pull/4245#issue-1529749481

File: src/contacts/ContactChat.ts [hunk 2]
```diff
@@ -292,25 +302,42 @@ export class ContactChat extends ContactStoreElement {
...
+      }
 
-    if (this.currentTicket) {
-      payload['ticket'] = this.currentTicket.uuid;
+      postJSON(`/api/v2/broadcasts.json`, payload)
+        .then(() => {
+          compose.currentChat = '';
+          compose.values = [];
+          compose.buttonDisabled = true;
+          this.refresh(true);
```
Comment (susanm74) [inline src/contacts/ContactChat.ts:334]: @ericnewcomer not sure why this isn't refreshing, but i assume once this does, it'll fix the "chat history not refreshing/displaying new msgs and attachments" issue that i'm encountering

File: src/contacts/ContactChat.ts [hunk 3]
```diff
@@ -48,8 +50,8 @@ export class ContactChat extends ContactStoreElement {
 
       .chatbox {
         padding: 1em;
-        background: #f2f2f2;
-        border-top: 3px solid #e1e1e1;
+        // background: #f2f2f2;
```
Comment (susanm74) [inline src/contacts/ContactChat.ts:53]: todo remove these styles as they aren't needed anymore

File: src/compose/Compose.ts [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+    filesToUpload.map(fileToUpload => {
+      this.uploadFile(fileToUpload);
+    });
+  }
+
+  private uploadFile(file: File): void {
+    console.log('uploadFile file', file);
+    this.uploading = true;
+
+    // todo remove for final PR
```
Comment (susanm74) [inline src/compose/Compose.ts:234]: todo remove this

File: src/compose/Compose.ts [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+            if (!chat.hasVisibleOptions()) {
+              this.handleSendEnter(e);
+              this.preventDefaults(e);
+            }
+          }
+        }}
+        placeholder="Type something here"
+      >
+      </temba-completion>
+      <temba-charcount text="Type something here"></temba-charcount>`;
```
Comment (susanm74) [inline src/compose/Compose.ts:374]: @ericnewcomer not sure why this isn't updating the count correctly when the temba completion text input value updates

File: src/compose/Compose.ts [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+        justify-content: space-between;
+      }
+      .highlight {
+        opacity: 0.5;
+      }
+      .items {
+        margin-bottom: 0.5em;
+      }
+
+      temba-completion {
```
Comment (susanm74) [inline src/compose/Compose.ts:40]: when the send button is clicked, there's a blue border that displays but then lingers, need to figure out if i'm missing temba-button override styles (similar to temba-completion) or whether this is due to the whole component needing a missing "refresh" fn

File: src/compose/Compose.ts [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+  endpoint = '/msgmedia/upload/';
+
+  @property({ type: String })
+  accept = ''; //e.g. ".xls,.xlsx"
+
+  @property({ type: Boolean, attribute: false })
+  uploading: boolean;
+
+  @property({ type: Array, attribute: false })
+  uploadErrors: string[] = [];
```
Comment (susanm74) [inline src/compose/Compose.ts:121]: is it possible to programmatically clear out the uploadErrors var upon ANY next action/click/event/etc? would that be where a blur or focusout comes in handy?

File: test/temba-compose.test.ts [hunk 8]
```diff
@@ -0,0 +1,231 @@
+import { assert, expect } from '@open-wc/testing';
```
Comment (susanm74) [inline test/temba-compose.test.ts:1]: todo create test assets

File: src/contacts/ContactChat.ts [hunk 9]
```diff
@@ -292,25 +302,42 @@ export class ContactChat extends ContactStoreElement {
...
+      postJSON(`/api/v2/broadcasts.json`, payload)
+        .then(() => {
+          compose.currentChat = '';
+          compose.values = [];
+          compose.buttonDisabled = true;
+          this.refresh(true);
+        })
+        .catch(err => {
+          console.error(err);
+          // todo display err msg below send button?
```
Comment (susanm74) [inline src/contacts/ContactChat.ts:338]: or we could display send button err msg in the same spot as the attachment error msgs, and we could make sure they still display even if/when the attachments are turned off, and only the chatbox is visible

File: src/compose/Compose.ts [hunk 10]
```diff
@@ -0,0 +1,457 @@
...
+          <label class="upload-label" for="upload-files">
+            <temba-icon
+              class="upload-icon"
+              name="${Icon.attachment}"
+              @click="${this.handleAddAttachments}"
+              clickable
+            ></temba-icon>
+            ${this.uploadErrors.length > 0
+              ? html` <div class="upload-error">
+                  ${this.uploadErrors.join(', ')}
```
Comment (susanm74) [inline src/compose/Compose.ts:432]: not sure why this join is only displaying the first or last string in the array, when it should be displaying the entire array of strings joined together and separated by commas

File: src/contacts/ContactChat.ts [hunk 11]
```diff
@@ -292,25 +302,39 @@ export class ContactChat extends ContactStoreElement {
...
+    console.log('handleSend evt', evt);
+    const buttonName = evt.detail.name;
+    if (buttonName === 'Send') {
+      const payload = {
+        contacts: [this.currentContact.uuid],
+      };
+      const compose = evt.currentTarget as Compose;
+      if (compose && (compose.currentChat || compose.values)) {
+        // todo confirm if this is needed now that we've got translations, request still requiring it
+        payload['text'] = compose.currentChat;
```
Comment (susanm74) [inline src/contacts/ContactChat.ts:315]: @rowanseymour my request is failing at the moment because it is complaining about a missing text field in the payload, so i've added one for now to maintain things working, but can you let me know if i need to remove this (or remove once something else gets merged?)
Screen Shot 2023-02-07 at 3 19 46 PM
Screen Shot 2023-02-07 at 3 19 58 PM

Comment (susanm74) [reply src/contacts/ContactChat.ts:304]: @rowanseymour also do you prefer "und" vs. "eng" for the translations object?

File: src/contacts/ContactChat.ts [hunk 12]
```diff
@@ -292,25 +285,38 @@ export class ContactChat extends ContactStoreElement {
...
-    const payload = {
-      contacts: [this.currentContact.uuid],
-      text: this.currentChat,
-    };
+  private handleSend(evt: CustomEvent) {
+    console.log('handleSend evt', evt);
+    const buttonName = evt.detail.name;
+    if (buttonName === 'Send') {
+      const payload = {
+        contacts: [this.currentContact.uuid],
```
Comment (norkans7) [inline src/contacts/ContactChat.ts:293]: Don't we want to allow sending to a group or multiple contacts here?
Would that work by making a custom function for `temba-button-clicked`?

Comment (susanm74) [reply src/contacts/ContactChat.ts:293]: @norkans7 yep, components will wire up their own temba-button-clicked event handler, this is just a case where the ContactChat component's handleSend event handler is meant to only send to one contact by design, but other components will wire up their own to send to a group or multiple contacts or do whatever else as needed - hope that answers your question!

File: src/compose/Compose.ts (hunk 10 above)
Comment (ericnewcomer) [reply src/compose/Compose.ts]: If I recall, pretty sure this was things not re-rendering since the array reference didn't change. Fix is to request an update after modifying the array.

Comment (susanm74) [reply src/compose/Compose.ts]: yes that did fix the issue! (sorry i just forgot to go through and mark this and a few other comments as moot/resolved!)

File: src/contacts/ContactChat.ts [hunk 13]
```diff
@@ -292,25 +281,70 @@ export class ContactChat extends ContactStoreElement {
...
+      };
+      const compose = evt.currentTarget as Compose;
+      if (compose) {
```
Comment (susanm74) [inline src/contacts/ContactChat.ts:292]: @rowanseymour just a final sanity check that it's ok add text and attachments directly to the root broadcast API request payload? this is currently working, but i wanted to confirm that this is ok, or whether i need to be putting everything into a translations object keyed by 'eng' or 'und' or something else? (at the moment it seems like the backend handles doing this, yeah?)

File: test/temba-compose.test.ts [hunk 14]
```diff
@@ -0,0 +1,301 @@
...
+      counter: true,
+    });
+    await assertScreenshot(
+      'compose/chatbox-counter-no-send-button',
+      getClip(compose)
+    );
+  });
+
+  //todo fix this test - send button should be enabled (works in demo)
+  it('chatbox with text', async () => {
```
Comment (susanm74) [inline test/temba-compose.test.ts:105]: @ericnewcomer not sure why this screenshot is wrong (button is still disabled, but should be enabled) - when i check the demo page and test manually, it's displaying correctly

File: test/temba-compose.test.ts [hunk 15]
```diff
@@ -0,0 +1,301 @@
...
+    await assertScreenshot(
+      'compose/chatbox-with-text-and-hit-enter',
+      getClip(compose)
+    );
+  });
+});
+
+describe('temba-compose attachments', () => {
+  //todo fix this test - send button should be enabled (works in demo)
+  it('attachments no send button', async () => {
```
Comment (susanm74) [inline test/temba-compose.test.ts:160]: @ericnewcomer not sure why this screenshot is wrong (paperclip should be visible but it's missing) - when i check the demo page and test manually, it's displaying correctly

File: src/contacts/ContactChat.ts (hunk 13 above)
Comment (rowanseymour) [reply src/contacts/ContactChat.ts:292]: text and attachments are separate fields on that endpoint so this is correct - we will however soon be pointing this toward /api/v2/msgs.json instead

File: test/temba-compose.test.ts [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+    );
+  });
+
+  it('attachments with success uploaded files', async () => {
+    const compose: Compose = await getCompose({
+      attachments: true,
+      button: true,
+    });
+
+    // todo
```
Comment (susanm74) [inline test/temba-compose.test.ts:187]: @ericnewcomer ok so mocking File and FileList is a bit tricky (which are the objects that result from selecting files to upload), here are some thoughts:
- option 1 (what i'm currently doing) - manually adding objects to compose.values and compose.errorValues - simple, but it's missing some of the logic we may want to be testing (such as the duplicate file filtering in the uploadFiles fn)
- option 2 - adding JSON to test-assets and populating values and errorValues via some sort of “list_endpoint” - this extra endpoint would only be needed/used for the purpose of testing so i wasn't sure if this approach is justified
- option 3 - adding real files to test-assets and populating values and errorValues via some sort of “fetch” method - this extra method would only be needed/used for the purpose of testing so i also wasn't sure if this approach is justified (also adding extra media files to the project could result in some bloat)

... some of the above options would also require switching the uploadFiles fn from private to public (and maybe others), but only for the purpose of testing, so again, wasn't sure if an approach that requires doing this makes the most sense or is justified... that said, some of these private fns do contain more of the "meat" of the logic that we may want to be testing, such as the such as the duplicate file filtering

File: test/temba-compose.test.ts [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+  //   // todo
+  // });
+});
+
+// todo attachments:
+// todo manual upload
+// todo drag and drop upload
+
+// todo contactchat:
+// handleSend success responses
```
Comment (susanm74) [inline test/temba-compose.test.ts:300]: @ericnewcomer should i be looking into using something like mockPost() to test that the contact chat handleSend success and error handling is working properly? and if not, do you recommend a certain approach for testing the contact chat handleSend response flavors?

File: src/contacts/ContactChat.ts (hunk 13 above)
Comment (susanm74) [reply src/contacts/ContactChat.ts:292]: ok cool, thanks!

File: test/temba-contact-chat.test.ts [hunk 18]
```diff
@@ -191,6 +195,178 @@ describe('temba-contact-chat - contact tests', () => {
...
+describe('temba-contact-chat - contact tests - handle send tests - text no attachments', () => {
+  // beforeEach(function () {
+  //   clock = useFakeTimers();
+  // });
+
+  // afterEach(function () {
+  //   clock.restore();
+  // });
+
+  it.only('with text no attachments - success response', async () => {
```
Comment (susanm74) [inline test/temba-contact-chat.test.ts:207]: @ericnewcomer couldn't quite get this working, so i tried a few different things, referring to temba-contact-fields and temba-modax as inspiration, but i think my confusion is coming down to:
- does a call to mockPost replace something like "send.click" or do we need both?
- how do we pass JSON to the mockPost endpoint given that it's regex and not a string? (currently i'm passing the path to the "request_payload" test-asset as part of the endpoint arg, but the other calls to mockPost only pass something more simple like a single uuid as part of the endpoint arg)
- how do we check the results of mockPost given that it's return type of void and not Promise<WebResponse>? (currently i'm passing the "response_body" JSON var as the body arg)
i think that once i get this first one working, i assume the other 7 should easily follow!

**Comment by susanm74** [comment]
@ericnewcomer also a few screenshot discrepancies...
- 2/33 screenshots the send button is disabled when it should be enabled... still not sure why these are still failing
      - attachments-with-success-files
      - chatbox-no-text-attachments-with-success-files
   - 6/33 screenshots the send button is enabled when it should be disabled... still not sure why these are still failing
      - chatbox-with-text-and-click-send
      - chatbox-with-text-and-hit-enter
      - chatbox-with-text-attachments-no-files-and-click-send
      - chatbox-with-text-attachments-no-files-and-hit-enter
      - chatbox-with-text-attachments-with-all-files-and-hit-enter
      - chatbox-with-text-attachments-with-success-files-and-hit-enter
   - 10/33 screenshots need more height (component or clip or otherwise) to fix elements getting cut off or being out of view in the screenshot... haven't totally worked out why setting a height in getCompose isn't being honored when other components seem to be able to successfully do this ex. getContactChat passes in a height of 500 and i believe this is being honored across all tests
      - attachments-with-all-files-and-click-send
      - attachments-with-all-files
      - chatbox-no-text-attachments-with-all-files-and-click-send
      - chatbox-no-text-attachments-with-all-files
      - chatbox-no-text-attachments-with-failure-files
      - chatbox-no-text-attachments-with-success-files
      - chatbox-with-text-attachments-with-all-files-and-click-send
      - chatbox-with-text-attachments-with-failure-files
      - chatbox-with-text-attachments-with-success-files-and-click-send
      - chatbox-with-text-attachments-with-success-files

**Comment by ericnewcomer** [comment]
I'm super confused on the state of things when I pulled and ran this. Let's discuss on the call.

**Comment by susanm74** [comment]
ok just marking in time that all of the above clarifications, notes, questions, etc resolved for now! (including the giant comment re: screenshots)

**Comment by susanm74** [comment]
@ericnewcomer i finished up everything else except for (and i'm currently investigating) the height discrepancy with the temba-charcount test case screenshots (4 total)... there's a getCharCount method which passes in a parentNode, which then gets passed into the fixture method... when i run the charcount tests from the main branch, the clientHeight is 24 for all test cases, but when i run the tests from my sl-new-compose-component branch, the clientHeight varies depending on the test case, will post the run output below, but i'm still not totally sure where this extra height is coming from (since the fixture method is part of an imported library and when i dug into the methods that it calls, there wasn't anything super obvious re: height)... i checked the other spots on the front-end where the charcount is used, and nothing seems amiss, so i'm wondering if there's anything glaring or obvious from your side?

**Comment by susanm74** [comment]
CharCount test run - main branch...

susanmlister@Susans-MacBook-Pro temba-components % yarn test test/temba-charcount.test.ts
Debugger listening on ws://127.0.0.1:53682/f1a503f2-3196-4526-beb8-b7187f25bd25
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
yarn run v1.22.19
$ wtr --node-resolve --coverage test/temba-charcount.test.ts
Debugger listening on ws://127.0.0.1:53691/f3e80c52-5414-45af-adc8-52a66ec3f436
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.

test/temba-charcount.test.ts:

 🚧 Browser logs:
      getCharCount parentNode clientWidth 250
      getCharCount parentNode clientHeight 24
      getCharCount counter clientWidth 0
      getCharCount counter clientHeight 0
      getClip width 250
      getClip height 250
      newClip width 250
      newClip height 250
      getCharCount parentNode clientWidth 250
      getCharCount parentNode clientHeight 24
      getCharCount counter clientWidth 0
      getCharCount counter clientHeight 0
      getClip width 250
      getClip height 250
      newClip width 250
      newClip height 250
      getCharCount parentNode clientWidth 250
      getCharCount parentNode clientHeight 24
      getCharCount counter clientWidth 0
      getCharCount counter clientHeight 0
      getClip width 250
      getClip height 250
      newClip width 250
      newClip height 250
      getCharCount parentNode clientWidth 250
      getCharCount parentNode clientHeight 24
      getCharCount counter clientWidth 0
      getCharCount counter clientHeight 0
      getClip width 250
      getClip height 250
      newClip width 250
      newClip height 250
      getCharCount parentNode clientWidth 250
      getCharCount parentNode clientHeight 24
      getCharCount counter clientWidth 0
      getCharCount counter clientHeight 0
      getClip width 250
      getClip height 250
      newClip width 250
      newClip height 250

Chromium: |██████████████████████████████| 1/1 test files | 5 passed, 0 failed

Code coverage: 48.15 %
View full coverage report at coverage/lcov-report/index.html

Finished running tests in 6.1s, all tests passed! 🎉

Waiting for the debugger to disconnect...
✨  Done in 7.28s.
Waiting for the debugger to disconnect...
susanmlister@Susans-MacBook-Pro temba-components %

**Comment by susanm74** [comment]
CharCount test run - sl-new-compose-component branch...

susanmlister@Susans-MacBook-Pro temba-components % yarn test test/temba-charcount.test.ts
Debugger listening on ws://127.0.0.1:54364/e48d81e9-8d02-4872-ab4b-0fdb606288d2
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
yarn run v1.22.19
$ wtr --node-resolve --coverage test/temba-charcount.test.ts
Debugger listening on ws://127.0.0.1:54371/abd8ca4b-2ef1-4954-bd13-a8c060b139c4
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.

test/temba-charcount.test.ts:

 🚧 Browser logs:
      getCharCount parentNode clientWidth 250
      getCharCount parentNode clientHeight 78
      getCharCount counter clientWidth 0
      getCharCount counter clientHeight 0
      getClip width 250
      getClip height 250
      newClip width 250
      newClip height 250
      getCharCount parentNode clientWidth 250
      getCharCount parentNode clientHeight 127
      getCharCount counter clientWidth 0
      getCharCount counter clientHeight 0
      getClip width 250
      getClip height 250
      newClip width 250
      newClip height 250
      getCharCount parentNode clientWidth 250
      getCharCount parentNode clientHeight 162
      getCharCount counter clientWidth 0
      getCharCount counter clientHeight 0
      getClip width 250
      getClip height 250
      newClip width 250
      newClip height 250
      getCharCount parentNode clientWidth 250
      getCharCount parentNode clientHeight 211
      getCharCount counter clientWidth 0
      getCharCount counter clientHeight 0
      getClip width 250
      getClip height 250
      newClip width 250
      newClip height 250
      getCharCount parentNode clientWidth 250
      getCharCount parentNode clientHeight 211
      getCharCount counter clientWidth 0
      getCharCount counter clientHeight 0
      getClip width 250
      getClip height 250
      newClip width 250
      newClip height 250

 ❌ temba-charcount > counts plain text
      Error: Screenshot was not the right size Expected 540x88 but got 540x196 
      /Users/susanmlister/Projects/textit/temba-components/screenshots/test/counter/text.png
      /Users/susanmlister/Projects/textit/temba-components/screenshots/truth/counter/text.png
      /Users/susanmlister/Projects/textit/temba-components/screenshots
        at assertScreenshot (test/utils.test.ts:186:12)
        at async o.<anonymous> (test/temba-charcount.test.ts:22:4)

 ❌ temba-charcount > counts variables
      Error: Screenshot was not the right size Expected 540x88 but got 540x294 
      /Users/susanmlister/Projects/textit/temba-components/screenshots/test/counter/variable.png
      /Users/susanmlister/Projects/textit/temba-components/screenshots/truth/counter/variable.png
      /Users/susanmlister/Projects/textit/temba-components/screenshots
        at assertScreenshot (test/utils.test.ts:186:12)
        at async o.<anonymous> (test/temba-charcount.test.ts:30:4)

 ❌ temba-charcount > counts unicode
      Error: Screenshot was not the right size Expected 540x88 but got 540x364 
      /Users/susanmlister/Projects/textit/temba-components/screenshots/test/counter/unicode.png
      /Users/susanmlister/Projects/textit/temba-components/screenshots/truth/counter/unicode.png
      /Users/susanmlister/Projects/textit/temba-components/screenshots
        at assertScreenshot (test/utils.test.ts:186:12)
        at async o.<anonymous> (test/temba-charcount.test.ts:37:4)

 ❌ temba-charcount > counts unicode with variables
      Error: Screenshot was not the right size Expected 540x88 but got 540x462 
      /Users/susanmlister/Projects/textit/temba-components/screenshots/test/counter/unicode-variables.png
      /Users/susanmlister/Projects/textit/temba-components/screenshots/truth/counter/unicode-variables.png
      /Users/susanmlister/Projects/textit/temba-components/screenshots
        at assertScreenshot (test/utils.test.ts:186:12)
        at async o.<anonymous> (test/temba-charcount.test.ts:44:4)

Chromium: |██████████████████████████████| 1/1 test files | 1 passed, 4 failed

Code coverage: 47.43 %
View full coverage report at coverage/lcov-report/index.html

Finished running tests in 5.8s with 4 failed tests.

Waiting for the debugger to disconnect...
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Waiting for the debugger to disconnect...
susanmlister@Susans-MacBook-Pro temba-components %

**Comment by susanm74** [comment]
@ericnewcomer @rowanseymour @norkans7 barring a minor discrepancy with 5 charcount test failures due to a pixel mismatch caused by some extra whitespace below the component, would love to get your eyes/review on this if/when you get a chance!

**Comment by ericnewcomer** [comment]
I see the char count errors, but I also see:

```
 ❌ temba-contact-chat - contact tests - handle send tests - text no attachments > with text no attachments - success response
      AssertionError: expected 'sà-wàd-dee!' to equal ''
      + expected - actual
      -sà-wàd-dee!
      at o.<anonymous> (test/temba-contact-chat.test.ts:243:32)
```

When I looked to investigate, the file is filled with so much commented out code, things don't seem close for review at all. 

...and I needed to comment out so many console logs to even start looking at things. To be totally honest, I'm not sure the current state of this, or what exactly you are asking from me in review.

**Comment by ericnewcomer** [comment]
Please get this cleaned up and ready for review.

* No commented out tests
* No `.only` tests
* Tests passing

I also feel the level of test scenarios is maybe a bit over what is reasonable. We need to strike a better balance over covering the spirit of the component and maintainability of the test suite.

If you can't get the tests to fully pass, then get everything else ready / checked in and let me know and I'll pull it down and take a look.

**Comment by susanm74** [comment]
@ericnewcomer 
- No commented out tests - done
- No .only tests - done
- Tests passing - i'm not 100% sure why so many tests were failing (~100!), but it looks like things possibly got a bit wonky due to a combo of some yarn.lock updates and a missing check for null needed in the compose code - i've got things back down to 1 test still failing locally for me in contact chat due to a timeout:
`❌ temba-contact-chat - ticket tests > show history and show chatbox if contact is active and ticket is open
      Error: waitUntil timed out after 1000ms
        at node_modules/@open-wc/testing-helpers/src/helpers.js:164:14`
   - i've stripped this test (and a few others) down to just an assertScreenshot, to eliminate some redundancy and anything that might be needlessly slowing the test down, but it's still failing - possible solutions include re-working the test (although it is very simple at this point and structured very much the same as others like it that are passing), increase the timeout for just this one test, get rid of this test if you think it's overkill...

and then yeah, i didn't mean to misrepresent this PR, i thought keeping it in draft mode was the way to signal that i wasn't looking for a final review, but i should've definitely been much more clear that i was just trying to get feedback on the code when it's the case that a) things are working on the frontend, but b) i'm trying to close in on test anomalies/failures... again, did not mean to cause frustration or confusion when requesting a review, just wanted to avoid the problems that can arise from a "big reveal" by not asking for feedback earlier in the process... but i realize that's what happened, and the above items you highlighted were problematic and should've been taken care of before requesting a review

and then yeah, definitely down to use this PR as a way to better understand how much of the testing burden is expected to be invested in upfront (vs starting with a more concise baseline and adding more as we go if needed) so that there's enough coverage when things are ready to test/ship, but not so much that it feels like it's dragging things out

**Comment by susanm74** [comment]
@ericnewcomer ok just got things fully updated with main, and looks like i'm now seeing 3 test failures total - 1 contact chat failure (the one i copy/pasted above), and 2 more contact history failures:
 ❌ temba-contact-history > renders history
      AssertionError: expected 0 to equal 549
      + expected - actual
      -0
      +549
      at o.<anonymous> (test/temba-contact-history.test.ts:76:19)

 ❌ temba-contact-history > expands event groups
      TypeError: Cannot read properties of null (reading 'click')
        at o.<anonymous> (test/temba-contact-history.test.ts:97:12)

so will look into those 3 and see if i can get things fully squared away with this PR!

**Comment by susanm74** [comment]
@ericnewcomer ok 3 tests still failing 😔 i spent some time today trying to track down the issue with each one (the two contact history test cases might be related, but the other contact chat test case is not), but couldn't quite close in on them (same 3 as above) - would be great if we can look at these together tonight/tomorrow, and hopefully resolve 'em , as it would get this PR pretty dang close if not ready to shift out of draft review and into regular/final review

**Comment by susanm74** [comment]
@ericnewcomer

- finished the chatbox updates re: the word break line wrapping issue (was also needed in contact history), added a few "long text" test cases, and regenerated all the screenshots that needed one

- but even with the updated predicate, i'm still seeing this one test failing on the server, it's also still failing for me locally - was it passing locally for you?
   - ❌ temba-contact-chat - ticket tests > show history and show chatbox if contact is active and ticket is open
      Error: waitUntil timed out after 1000ms
        at node_modules/@open-wc/testing-helpers/src/helpers.js:164:14

   - could you explain a bit more how these scrollTop predicate values are calculated given a component test width and height of 500x500? looks like each test case has a different scrollTop value x, and then a different value y is used in the predicate of each test case... is there any way to dynamically calculate these values on the fly? i didn't quite get to a point today to be able to determine if this is possible or how much work it would be, but if you can confirm, i'd be happy to do this as part of this PR or as a smaller follow-up PR

- will keep this in draft mode for now, but could you take another look at the test above? and then if you're ready to give the PR an overall prelim review, things are in a much better spot now

File: test/temba-contact-chat.test.ts [hunk 19]
```diff
@@ -501,13 +501,22 @@ describe('temba-contact-chat - ticket tests', () => {
...
       getClip(chat),
       {
         clock: clock,
         predicate: () => {
-          return chat.getContactHistory().getEventsPane().scrollTop === 1024;
+          const currentScroll = chat
+            .getContactHistory()
+            .getEventsPane().scrollTop;
+          if (currentScroll !== 0 && currentScroll === lastScroll) {
+            return true;
```
Comment (ericnewcomer) [inline test/temba-contact-chat.test.ts:517]: @susanm74 instead I just made this predicate rely on a specific scrollTop value (which was different because our chatbox height changed and could change similarly in the future). This should hopefully be more future proof.

File: src/compose/Compose.ts [hunk 20]
```diff
@@ -258,14 +287,12 @@ export class Compose extends FormElement {
   }
 
   private highlight(evt: DragEvent): void {
-    const dragAndDropZone = evt.target as HTMLDivElement;
-    dragAndDropZone.classList.add('highlight');
+    this.pendingDrop = true;
```
Comment (ericnewcomer) [inline src/compose/Compose.ts:290]: @susanm74 we don't want to manipulate classList directly, instead it should be declarative by setting a property and then using that property in the render methods to set the class.

File: src/compose/Compose.ts [hunk 21]
```diff
@@ -448,56 +477,55 @@ export class Compose extends FormElement {
...
-                      (${formatFileSize(0, 0)}) - Attachment failed</span
-                    >
-                  </div>
-                </div>`;
-              })}
-            </div`
+                </div>
+              </div>`;
+            })}
+          </div>`
```
Comment (ericnewcomer) [inline src/compose/Compose.ts:528]: All I changed here was adding the last `>` that was missing. Looks like we have formatters that don't match though. We need to make sure things are formatted the same way on a commit hook (I thought we had that) so our diffs aren't chatty.

File: src/compose/Compose.ts [hunk 22]
```diff
@@ -413,12 +440,14 @@ export class Compose extends FormElement {
   public render(): TemplateResult {
     return html`
       <div
-        class="container"
+        class=${getClasses({ container: true, highlight: this.pendingDrop })}
```
Comment (ericnewcomer) [inline src/compose/Compose.ts:443]: Here is where I used the property declaratively

File: src/compose/Compose.ts [hunk 23]
```diff
@@ -42,12 +44,36 @@ export class Compose extends FormElement {
...
+        right: 0;
+        background: rgba(210, 243, 184, 0.8);
+        border-radius: var(--curvature-widget);
+        margin: -0.5em;
+        padding: 0.5em;
+        transition: opacity ease-in-out var(--transition-speed);
+        display: flex;
+        align-items: center;
+        text-align: center;
+      }
```
Comment (ericnewcomer) [inline src/compose/Compose.ts:66]: Switched the drop treatment to be a single div that overlays the entire component with an semi-transparent background.

File: src/charcount/CharCount.ts [hunk 24]
```diff
@@ -51,10 +51,11 @@ export class CharCount extends RapidElement {
...
           var(--transition-speed);
         visibility: hidden;
-        margin-top: 5px;
-        right: 0px;
+        margin-top: var(--temba-charcount-summary-margin-top);
+        right: var(--temba-charcount-summary-right);
+        bottom: var(--temba-charcount-summary-bottom);
         text-align: left;
-        position: absolute;
+        position: var(--temba-charcount-summary-position);
```
Comment (ericnewcomer) [inline src/charcount/CharCount.ts:58]: We'll need to remember to add these to temba-components.css on rapid

File: src/compose/Compose.ts (hunk 20 above)
Comment (susanm74) [reply src/compose/Compose.ts:290]: 👍

File: src/compose/Compose.ts (hunk 21 above)
Comment (susanm74) [reply src/compose/Compose.ts:528]: cool, sounds like the commit hook should fix that, but let me know if there's any local format settings that i can change or share in the meantime

File: src/compose/Compose.ts (hunk 23 above)
Comment (susanm74) [reply src/compose/Compose.ts:66]: 👍

File: src/charcount/CharCount.ts (hunk 24 above)
Comment (susanm74) [reply src/charcount/CharCount.ts:58]: i can go ahead and create a PR for that if there isn't one already

File: src/compose/Compose.ts (hunk 22 above)
Comment (susanm74) [reply src/compose/Compose.ts:443]: 👍

File: src/charcount/CharCount.ts (hunk 24 above)
Comment (susanm74) [reply src/charcount/CharCount.ts:58]: ok looks like that's already done per https://github.com/nyaruka/rapidpro/pull/4371/files
